### PR TITLE
SONARJAVA-1434 RSPEC-3417 Disallowed dependencies should not be used (simplified parameters)

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/test/java/org/sonar/java/checks/maven/helpers/MavenDependencyVersionMatcherTest.java [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+package org.sonar.java.checks.maven.helpers;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.sonar.maven.model.LocatedAttribute;
+import org.sonar.maven.model.maven2.Dependency;
+
+import static org.fest.assertions.Assertions.assertThat;
+
+public class MavenDependencyVersionMatcherTest {
```
Comment (benzonico) [inline java-checks/src/test/java/org/sonar/java/checks/maven/helpers/MavenDependencyVersionMatcherTest.java]: This test class is really long. 
I didn't investigate in deep but I'm pretty sure there are a lot of redundant tests.
Moreover : you can factorize things : 
```
setDependency(StringVar); 
assertThat(matcher.matches(dependency)).isTrue();
```
matcher can be made into a field and you can then declare assertMatch and assertNotMatch methods that would take a string parameter.

File: java-checks/src/main/java/org/sonar/java/checks/maven/helpers/MavenDependencyVersionMatcher.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+   *     <li>"<code>*-3.1</code>" : any version before version 3.1</li>
+   *   </ul>
+   *   <li>empty (when version is not relevant): ""</li>
+   * </ul>
+   *
+   * @param version the version expected for the dependency
+   * @return
+   */
+  public static MavenDependencyVersionMatcher fromString(String version) {
+    if (version.isEmpty() || isWildCard(version)) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/maven/helpers/MavenDependencyVersionMatcher.java]: What if version is null or blank ? version = "   ";

File: java-checks/src/main/java/org/sonar/java/checks/maven/helpers/MavenDependencyAbstractMatcher.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+package org.sonar.java.checks.maven.helpers;
+
+import org.sonar.maven.model.LocatedAttribute;
+import org.sonar.maven.model.maven2.Dependency;
+
+import javax.annotation.Nullable;
+
+import java.util.regex.Pattern;
+
+public abstract class MavenDependencyAbstractMatcher {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/maven/helpers/MavenDependencyAbstractMatcher.java]: Hierarchy of classes seems odd in this case as version and name of depednency have nothin in common and should not match on same parameter

File: java-checks/src/main/java/org/sonar/java/checks/maven/DisallowedDependenciesCheck.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    description = "Dependency version pattern or dash-delimited range. Leave blank for all versions. "
+      + "E.G. <code>1.3.*</code>, <code>1.0-3.1</code>, <code>1.0-*</code> or <code>*-3.1</code>")
+  private String version = "";
+
+  @Override
+  public void scanFile(MavenFileScannerContext context) {
+
+    MavenDependencyMatcher matcher;
+    try {
+      matcher = MavenDependencyMatcher.withName(dependencyName).withVersion(version);
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/maven/DisallowedDependenciesCheck.java]: Seems odd to reconstruct this object for every single file... if matcher is stateless for matching (which it should) then it should be built once.

File: java-checks/src/main/java/org/sonar/java/checks/maven/DisallowedDependenciesCheck.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      throw new IllegalArgumentException("[" + KEY + "] Unable to build matchers from provided dependency name: " + dependencyName, e);
+    }
+    for (Dependency dependency : new MavenDependencyCollector(context.getMavenProject()).allDependencies()) {
+      if (matcher.matches(dependency)) {
+        context.reportIssue(this, dependency, "Remove this forbidden dependency.");
+      }
+    }
+  }
+
+  public void setDependencyName(String dependencyName) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/maven/DisallowedDependenciesCheck.java]: no need for setter, set fields public.

File: java-checks/src/test/java/org/sonar/java/checks/maven/helpers/LocatedAttributeMatcherTest.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.maven.helpers;
+
+public class LocatedAttributeMatcherTest {
```
Comment (m-g-sonar) [inline java-checks/src/test/java/org/sonar/java/checks/maven/helpers/LocatedAttributeMatcherTest.java]: will be removed.

File: java-checks/src/main/java/org/sonar/java/checks/maven/helpers/LocatedTreeMatcher.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import javax.annotation.Nullable;
+
+public interface LocatedTreeMatcher {
+
+  /**
+   * Checks that a located tree matches the matcher.
+   * @param tree The tree to check
+   * @return true if the tree matches the matcher, false otherwise
+   */
+  public boolean matches(@Nullable LocatedTree tree);
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/maven/helpers/LocatedTreeMatcher.java]: MINOR "public" is redundant in this context. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2333)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/maven/helpers/LocatedTreeMatcher.java (hunk 7 above)
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/maven/helpers/LocatedTreeMatcher.java]: available also for the classes.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/maven/helpers/LocatedAttributeMatcher.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+ */
+package org.sonar.java.checks.maven.helpers;
+
+import org.sonar.maven.model.LocatedAttribute;
+
+import javax.annotation.Nullable;
+
+public interface LocatedAttributeMatcher {
+
+  public boolean matches(@Nullable LocatedAttribute tree);
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/maven/helpers/LocatedAttributeMatcher.java]: MINOR "public" is redundant in this context. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2333)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/maven/helpers/LocatedAttributeMatcher.java (hunk 8 above)
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/maven/helpers/LocatedAttributeMatcher.java]: to be fixed, and this is also available for the inner class in this fiel.

File: java-checks/src/main/java/org/sonar/java/checks/maven/helpers/MavenDependencyMatcher.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+   * @param dependencyName the dependency pattern. Expected format: "<code>[groupId]:[artifactId]</code>",
+   *  with following options for both [groupId] and [artifactId]:
+   *   <ul>
+   *     <li>Wildcard (i.e. "<code>*:myArtifactId</code>" or "<code>myGroupId:*</code>")</li>
+   *     <li>Regular expression (i.e. "<code>org.sonar.*:*java</code>")</li>
+   *   </ul>
+   *
+   * @param version the version. Expected formats:
+   *   <ul>
+   *     <li>Enpty string (i.e. "")</li>
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/maven/helpers/MavenDependencyMatcher.java]: typo

**Pull request merged**